### PR TITLE
Move Package.Dependency.Constraint to Dune_lang

### PR DESCRIPTION
**Pull request by gridbugs**
This will let us use the constraint language within lockfiles which we need in order for build and install commands be conditional on opam filters.

### CODE REVIEW

File: src/dune_lang/package_constraint.mli [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+        (** A variable name such as :version or :with-test (the leading ":" is
+            excluded from the string) *)
+end
+
+type t =
+  | Bvar of Var.t
+  | Uop of Op.t * Var.t
+  | Bop of Op.t * Var.t * Var.t
+  | And of t list
+  | Or of t list
```
Comment (Alizter) [inline src/dune_lang/package_constraint.mli:27]: Can you add comments for these?

**Review by Leonidas-from-XIV — COMMENTED** [review]
I think this PR makes a good deal of sense, some comments that popped into my head while I was reviewing below.

File: src/dune_lang/package_constraint.ml [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    | Var of string
+
+  let encode = function
+    | Literal v -> Dune_sexp.Encoder.string v
+    | Var v -> Dune_sexp.Encoder.string (":" ^ v)
+
+  let decode =
+    let open Dune_sexp.Decoder in
+    let+ s = string in
+    if String.is_prefix s ~prefix:":" then Var (String.drop s 1) else Literal s
```
[the remark this answers is not in the document]
Comment (gridbugs) [reply src/dune_lang/package_constraint.ml:41]: I'll do that in a separate PR

File: src/dune_rules/package.mli [hunk 3]
```diff
@@ -33,27 +33,27 @@ module Id : sig
...
-      | Gt
-      | Lt
-      | Neq
-
-    val to_relop : t -> OpamParserTypes.FullPos.relop
-  end
-
   module Constraint : sig
+    module Op : sig
+      type t = Dune_lang.Package_constraint.Op.t =
```
[the remark this answers is not in the document]
Comment (gridbugs) [reply src/dune_rules/package.mli:38]: also `Var.t`

File: src/dune_lang/package_constraint.mli (hunk 1 above)
Comment (gridbugs) [reply src/dune_lang/package_constraint.mli]: Done

**Comment by gridbugs** [comment]
In the process of documenting the different `Constraint.t` constructors I noticed it was possible to have a `Bvar` which contained a string literal. Even though the parser was incapable of producing such a value, the fact that it could be represented made the API harder for me to understand. I've changed it so that only variables can be passed to the `Bvar` constructor and made a clear distinction between the `Variable.t` and `Value.t` types (previously string literals were in a type called `Var.t` which made them seem like variables and I've always found that confusing).

**Comment by rgrinberg** [comment]
The change in the description looks fine, but I see you've started pushing some additional work on top. Was that intentional?

File: src/dune_lang/package_constraint.mli [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    | Variable of Variable.t
+end
+
+(** A constraint for use in package definitions, equivalent to the filter
+    language used in opam package definitions. All expressions in this language
+    are booleans. String literal values can only appear as the argument to an
+    operator which returns a boolean. Variables are dynamically typed and can
+    represent strings and booleans. If a variable appears in a position where a
+    boolean is expected it will be assumed to represent a boolean. *)
+type t =
```
Comment (Alizter) [inline src/dune_lang/package_constraint.mli:29]: Dumb question: Is our blang really that different from this?

**Comment by gridbugs** [comment]
> The change in the description looks fine, but I see you've started pushing some additional work on top. Was that intentional?

Oops I accidentally pushed some additional work here. I just removed it.

File: src/dune_lang/package_constraint.mli (hunk 4 above)
Comment (gridbugs) [reply src/dune_lang/package_constraint.mli:29]: Good question. I don't know off the top of my head. Of course they both have similar functions though the package constraint language has the added requirement of being translatable to the opam filter language. At some point soon I'll be looking at actually evaluating conditions (currently we only use these to generate opam files) and at that point I'll see whether it makes sense to just replace this with blang.

**Pull request merged**